test/shared/src/main/scala/zio/test/TestReporter.scala

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import zio.{ URIO, ZIO }
+import zio.duration.Duration
+
+/**
+ * A `TestReporter[L]` is capable of reporting test results annotated with
+ * labels `L`.
+ */
+trait TestReporter[-L] {
+  def apply[E, S](duration: Duration, executedSpec: ExecutedSpec[L, E, S]): URIO[TestLogger, Unit]
```

Comment:
If we make `TestReporter` polymorphic in `E` and `S`, we won't be able to have reporters that do specific things with `E` or `S`. Since that could be useful, we may want to stick the type parameters in `TestReporter[-L, +E, +S]`, so some types of test reporters can take advantage of their knowledge of concrete error and status types.